cmd/signature-v4-utils.go

```diff
@@ -157,3 +157,25 @@ func extractSignedHeaders(signedHeaders []string, reqHeaders http.Header) (http.
 
 	return extractedSignedHeaders, ErrNone
 }
+
+// Sequential spaces regular expression
+var seqSpaces = regexp.MustCompile("  +")
+
+// Trim leading and trailing spaces and replace sequential spaces with one space, following Trimall()
+// in http://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html
+func sigV4TrimAll(input string) string {
+	// Remove all whitespaces first
+	cleanWhiteSpaces := func(r rune) rune {
+		switch r {
+		case '\t', '\n', '\u000b', '\r', '\f':
+			return ' '
+		}
+		return r
+	}
+	input = strings.Map(cleanWhiteSpaces, input)
+	// Clean leading and trailing spaces
+	input = strings.TrimSpace(input)
+	// Compress adjacent spaces to one space
+	replaced := seqSpaces.ReplaceAllString(input, " ")
```

Comment:
I think this function can be simplified to just one line:

```
return strings.Join(strings.Fields(input), " ")
```
Just found https://golang.org/pkg/strings/#Fields - splits string after compressing consecutive whitespace. Uses unicode's definition of whitespace: https://golang.org/pkg/unicode/#IsSpace https://en.wikipedia.org/wiki/Whitespace_character#Unicode